### PR TITLE
TimeSeries: skip multi-frame alignment when all timestamps match

### PR DESCRIPTION
**Pull request by leeoniya**
does what it says on the tin.

when all frames have matching timestamps, this reduces CPU time of `joinDataFrames()` from 400ms -> 17ms for panels with 1,600,000 datapoints: https://github.com/grafana/grafana/issues/62005#issuecomment-1411660357.

should do a bit more testing, we might rely on `joinDataFrames()` to sort and null-fill in some cases, even for a single frame.

an easy way to test this is to watch the console while playing with `gdev-testdata` and multiple frames/series:

https://github.com/grafana/grafana/assets/43234/6d7100a7-b721-40c4-9aa5-8359003223ec

### CODE REVIEW

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts [hunk 1]
```diff
@@ -149,10 +149,77 @@ export function joinDataFrames(options: JoinOptions): DataFrame | undefined {
...
+                      frameIndex,
+                      fieldIndex,
+                    }
+                  };
+                }
+
+                return fieldCopy;
+              }).filter((field) => field != null)
+            ),
+          ],
```
Comment (mckn) [inline packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:212]: nit: maybe worth breaking this out into its own function for readability?

**Review by mckn — APPROVED** [review]
LGTM!

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts (hunk 1 above)
Comment (leeoniya) [reply packages/grafana-data/src/transformations/transformers/joinDataFrames.ts]: done!

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts [hunk 2]
```diff
@@ -148,10 +227,16 @@ export function joinDataFrames(options: JoinOptions): DataFrame | undefined {
     return frameCopy;
   }
 
+  let joinedFrame = cheapJoinDataFrames(options);
+
+  if (joinedFrame != null) {
+    console.log('did cheapJoin!');
```
Comment (leeoniya) [inline packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:233]: here for testing

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts [hunk 3]
```diff
@@ -82,159 +83,131 @@ export function maybeSortFrame(frame: DataFrame, fieldIdx: number) {
...
+    length: joinedData[0].length,
+    fields: allFields.map((f, index) => ({
       ...f,
-      values: joined[index],
+      values: joinedData[index],
     })),
   };
+
+  if (cheap) {
+    // console.log('cheap!');
```
Comment (baldm0mma) [inline packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:206]: console.log

Comment (leeoniya) [reply packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:206]: //

Comment (baldm0mma) [reply packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:206]: Listen here ya little...

**Comment by leeoniya** [comment]
bad bot! no donut!

**Pull request closed without merging**